src/string.js

```diff
@@ -173,6 +173,37 @@ export function trimStart(str) {
   return (str + '_').trim().slice(0, -1);
 }
 
+/**
+ * Wrapper around String.replace that handles asynchronous resolution.
+ * @param {string} str
+ * @param {RegExp} regex
+ * @param {Function} replacer
+ * @return {!Promise<string>}
+ */
+export function asyncStringReplace(str, regex, replacer) {
+  const stringBuilder = [];
+  let lastIndex = 0;
+  str.replace(regex, function() {
+    // String.prototype.replace will pass 3 to n number of arguments to the
+    // callback function based on how many capture groups the regex may or may
+    // not contain. We know that the match will always be first, and the
+    // index will always be second to last.
+    const match = arguments[0];
+    const matchIndex = arguments[arguments.length - 2];
+    stringBuilder.push(str.slice(lastIndex, matchIndex));
+    // Store the promise in it's eventual string position.
+    const replacementPromise =
+      typeof replacer === 'function'
+        ? replacer.apply(null, arguments)
+        : replacer;
```

Comment:
The type says this is always a function. How do we get to this else branch?

Reply:
I think that this is just a bad type on my part, was trying to keep it consistent with `string.replace` where the replacer can be a `{Function|string}`. e.g. test-string.js L170 in this PR.

Reply:
Then we need to check it earlier. We can't support the `string` replacer here, since we've already passed a function to `string.replace()`. But if you check before calling string replace, you can just `Promise.resolve(string.replace(regex, replacment))`.

Reply:
Sorry I don't think I understand the difference. We are only really using native `string.replace` to get the matches & offset? Can you give an example that breaks it?

Reply:
`'t123'.replace(/t(\d+)/g, '$1')`. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_string_as_a_parameter, the string `replacement` value is much more complicated then most realize. Let's either forbid a string, or support it properly.

Reply:
ah thanks, you are totally right. Moving the check sgtm.

Reply:
also @micajuine-ho can we add a test case or two using these "special replacement patterns" :)